streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java

```diff
@@ -16,78 +16,408 @@
-     * @param tasks                   the set of tasks to reassign if possible. Must already be assigned
-     *                                to a KafkaStreams client
+     * @param tasks                   the set of tasks to reassign if possible. Must already be assigned to a KafkaStreams client
      *
-     * @return a new map containing the mappings from KafkaStreamsAssignments updated with the default
-     *         rack-aware assignment for active tasks
+     * @return a new map containing the mappings from KafkaStreamsAssignments updated with the default rack-aware assignment for active tasks
      */
     public static Map<ProcessId, KafkaStreamsAssignment> optimizeRackAwareActiveTasks(
         final ApplicationState applicationState,
         final Map<ProcessId, KafkaStreamsAssignment> kafkaStreamsAssignments,
         final SortedSet<TaskId> tasks
     ) {
-        throw new UnsupportedOperationException("Not Implemented.");
+        if (tasks.isEmpty()) {
+            return kafkaStreamsAssignments;
+        }
+
+        if (!hasValidRackInformation(applicationState)) {
+            LOG.warn("Cannot optimize active tasks with invalid rack information.");
+            return kafkaStreamsAssignments;
+        }
+
+        final int crossRackTrafficCost = applicationState.assignmentConfigs().rackAwareTrafficCost();
+        final int nonOverlapCost = applicationState.assignmentConfigs().rackAwareNonOverlapCost();
+        final long currentCost = computeTaskCost(
+            applicationState.allTasks().stream().filter(taskInfo -> tasks.contains(taskInfo.id())).collect(
+                Collectors.toSet()),
+            applicationState.kafkaStreamsStates(false),
+            crossRackTrafficCost,
+            nonOverlapCost,
+            false,
+            false
+        );
+        LOG.info("Assignment before active task optimization has cost {}", currentCost);
+
+        final List<UUID> clientIds = kafkaStreamsAssignments.keySet().stream().map(ProcessId::id).collect(
+            Collectors.toList());
+        final Map<ProcessId, KafkaStreamsState> kafkaStreamsStates = applicationState.kafkaStreamsStates(false);
+        final Map<UUID, Optional<String>> clientRacks = kafkaStreamsStates.values().stream().collect(
+                Collectors.toMap(state -> state.processId().id(), KafkaStreamsState::rackId));
+        final Map<UUID, Set<TaskId>> previousTaskIdsByProcess = kafkaStreamsStates.values().stream().collect(Collectors.toMap(
+            state -> state.processId().id(),
+            KafkaStreamsState::previousActiveTasks
+        ));
+        final Map<TaskId, Set<TaskTopicPartition>> topicPartitionsByTaskId = applicationState.allTasks().stream()
+            .filter(taskInfo -> tasks.contains(taskInfo.id()))
+            .collect(Collectors.toMap(TaskInfo::id, TaskInfo::topicPartitions));
+
+        final List<TaskId> taskIds = new ArrayList<>(tasks);
```

Comment:
We're computing some things over and over again, for example we reconstruct this same `List<TaskId>` three times for the active task optimization. The `List<UUID` gets computed twice. These may be smallish sets but the assignor should still be sensitive to memory consumption and performance. And just from a review perspective, I'm getting a bit overwhelmed by how many/how often we're creating new variations of these data structures via `.stream()` 

It seems like most of the rack aware/graph methods have the same input parameters, so it makes sense to just construct them all at the top and then pass those around as needed. When we wait to do these Set->List or ProcessId->UUID conversions until right before they're needed, it's too hard to keep track of whether we may have already needed/created that same thing in an earlier method call or later in the same method, etc

Reply:
I ended up having to redo a bunch of the data structures as part of the KafkaStreamsAssignment POC ([this](https://github.com/apourchet/kafka/pull/1)) so I happened to address this comment as well. 

You can see in the POC PR -- there's not actually very much we need to compute if we just do it all up front